subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCrossBuildInMemoryCacheFactory.java

```diff
@@ -171,14 +173,14 @@ private static class DefaultCrossBuildInMemoryCache<K, V> extends AbstractCrossB
         private final Map<K, SoftReference<V>> allValues;
 
         public DefaultCrossBuildInMemoryCache(Map<K, SoftReference<V>> allValues) {
-            this.allValues = allValues;
+            this.allValues = synchronizedMap(allValues);
```

Comment:
synchronizedMap pretty much just sucks. 

I would vote to require (and document on this method's javadoc) that this cache _must_ accept a thread-safe map as a parameter. 

Then, we can either pass in a `ConcurrentHashMap` or a `synchronizedMap(WeakHashMap)` depending on the factory function.

Or better yet, do not allow users to pass in a map at all and only expose factory functions that pass in safe maps into this constructor

Reply:
> I would vote to require (and document on this method's javadoc) that this cache _must_ accept a thread-safe map as a parameter.

Yes, that sounds like an improvement